### DISCUSSION TITLE
Question About Importing Timelines at Runtime

### DISCUSSION THREAD

**Discussion by howeswarderbutcool** [dialogic-godot/dialogic#2200] [Q&A]
So I want to Import Timelines at Runtime from the `user://` path that are not being created by the same dialogic.
Or in another words make it possible that, people can share their Timelines with characters, variables and so on to other people who can import them at runtime. 

I tried this `Dialogic.start(path_to_timeline)` and it actually works but im getting an error on the variables and the character. 
Variable: 
```
E 0:00:07:0621   subsystem_expression.gd:36 @ execute_string(): Invalid named index 'user_name' for base type Object
  <C++-Fehler>   Condition "p_show_error" is true. Returning: Variant()
  <C++-Quelle>   core/math/expression.cpp:1507 @ execute()
  <Stacktrace>   subsystem_expression.gd:36 @ execute_string()
                 subsystem_variables.gd:92 @ get_variable()
                 subsystem_variables.gd:52 @ parse_variables()
                 subsystem_text.gd:88 @ parse_text()
                 event_text.gd:106 @ _execute()
```

Character: The image will just not show up. 

My guess it that these need to be imported in another way? if this is even possible. 
If so i would love to get some help.

I use Godot4 and Dialogic2 btw.

**Comment by Jowan-Spooner** [comment]
Hmm... This won't be super straight-forward. It will work for timelines if you use the direct path, but characters and variables won't work as easily. 

Variables are stored in a dictionary in the project settings. You might be able to update that dictionary dynamically from an external dict I guess. 

Characters mainly won't work, because the unique-identifier directory (basically a fancy lookup dictionary) will not take paths in user:// into account (only in res://). It might also be possible to merge two of these directories (if you want to simply export the directory from another project to make it easier.) *By the way the same problem of course applies to timelines if you want to use the unique identifieres (e.g. in a jump event).*
You might have to manually add those characters to the directory, which might look something like this:
```
var directory := DialogicResourceUtil.get_directory("dtl")
directory[CHARACTER_IDENTIFIER] = FILE_PATH
DialogicResourceUtil.set_directory("dtl", directory)
``` 
This assumes your users somehow provide this info by having exported/given you their character and timeline directories (which are also saved in the project settings!).

**My main concern would be how you would handle it if someone imports variables/characters that clash with existing characters/variables.**

**Reply by howeswarderbutcool** [reply]
Thanks for you're answer and sorry for the late response.
So actually I was hoping that there was a built in function for this but I guess I need a few more brain cells. 

I tried you're idea and i guess this could work but it need a bit more tinkering. the since the .dch file reads from a specified path the image(and maybe dependence on some more information i would need to change) . I couldn't get i to work but while trying i got two new ideas. 

1. maybe not the best result and actually not what i was planning to do but maybe works as a workaround.
  I predefine these variables and or characters and make some dummy files in my project. (not sure if the dummy files are necessary, but that is my first guess)
  I could try to read at runtime the external files(png, dch, dtl, and so on) and overwrite the internal. Will this work. idk, I hope so.
  and well for you're question. in this case this would be no problem.

2. While testing i found a few constructors for the Dialogic characters. maybe i could create the characters via code. But idk if this is possible and it seems kinda complicated.
